GoogleDataTransportCCTSupport/GoogleDataTransportCCTSupport/Classes/GDTCCTNanopbHelpers.m

```diff
@@ -0,0 +1,159 @@
+  logRequest.has_client_info = 1;
+  logRequest.log_event = malloc(sizeof(gdt_cct_LogEvent) * logSet.count);
+  int i = 0;
+  for (GDTStoredEvent *log in logSet) {
+    gdt_cct_LogEvent logEvent = GDTCCTConstructLogEvent(log);
+    logRequest.log_event[i] = logEvent;
+    i++;
+  }
+  logRequest.log_event_count = (pb_size_t)logSet.count;
+
+  return logRequest;
+}
+
+gdt_cct_LogEvent GDTCCTConstructLogEvent(GDTStoredEvent *event) {
+  gdt_cct_LogEvent logEvent = gdt_cct_LogEvent_init_zero;
+  logEvent.event_time_ms = event.clockSnapshot.timeMillis;
+  logEvent.has_event_time_ms = 1;
+  logEvent.event_uptime_ms = event.clockSnapshot.uptime;
+  logEvent.has_event_uptime_ms = 1;
+  logEvent.timezone_offset_seconds = event.clockSnapshot.timezoneOffsetSeconds;
+  logEvent.has_timezone_offset_seconds = 1;
+  // TODO: Read network_connection_info from the custom params dict.
+
+  NSError *error;
+  NSData *extensionBytes = [NSData dataWithContentsOfURL:event.eventFileURL options:0 error:&error];
+  NSCAssert(error == nil, @"There was an error reading extension bytes from disk: %@", error);
+  logEvent.source_extension = GDTCCTEncodeData(extensionBytes);  // read bytes from the file.
+  return logEvent;
+}
+
+gdt_cct_ClientInfo GDTCCTConstructClientInfo() {
+  gdt_cct_ClientInfo clientInfo = gdt_cct_ClientInfo_init_zero;
+  clientInfo.client_type = gdt_cct_ClientInfo_ClientType_IOS_FIREBASE;
+  clientInfo.has_client_type = 1;
+  clientInfo.ios_client_info = GDTCCTConstructiOSClientInfo();
+  clientInfo.has_ios_client_info = 1;
+  return clientInfo;
+}
+
+gdt_cct_IosClientInfo GDTCCTConstructiOSClientInfo() {
+  gdt_cct_IosClientInfo iOSClientInfo = gdt_cct_IosClientInfo_init_zero;
+  UIDevice *device = [UIDevice currentDevice];
+  NSBundle *bundle = [NSBundle mainBundle];
+  NSLocale *locale = [NSLocale currentLocale];
+  iOSClientInfo.os_full_version = GDTCCTEncodeString(device.systemVersion);
+  NSArray *versionComponents = [device.systemVersion componentsSeparatedByString:@"."];
+  iOSClientInfo.os_major_version = GDTCCTEncodeString(versionComponents[0]);
+  NSString *version = [bundle objectForInfoDictionaryKey:(NSString *)kCFBundleVersionKey];
+  if (version) {
+    iOSClientInfo.application_build = GDTCCTEncodeString(version);
+  }
+  iOSClientInfo.country = GDTCCTEncodeString([locale objectForKey:NSLocaleCountryCode]);
+  iOSClientInfo.model = GDTCCTEncodeString(device.model);
+  NSString *languageCode = bundle.preferredLocalizations.firstObject;
+  iOSClientInfo.language_code =
+      languageCode ? GDTCCTEncodeString(languageCode) : GDTCCTEncodeString(@"en");
+  iOSClientInfo.application_bundle_id = GDTCCTEncodeString(bundle.bundleIdentifier);
+  return iOSClientInfo;
+}
+
+#pragma mark - CCT Object decoders
+
+gdt_cct_LogResponse GDTCCTDecodeLogResponse(NSData *data) {
+  gdt_cct_LogResponse response;
+  pb_istream_t istream = pb_istream_from_buffer([data bytes], [data length]);
+  pb_decode(&istream, gdt_cct_LogResponse_fields, &response);
```

Comment:
You should check the return value of pb_decode() to ensure it worked.

Reply:
Done.